notes-0319-inductive-types.tex

```diff
@@ -0,0 +1,228 @@
+
+\begin{remark}
+  While it was perfectly valid for us to have a constructor of type $\N \rightarrow \N$ when defining an inductive type, not all types make sense as the type of a constructor.
+  For example if we want to define a type $C$, a constructor of type $(C \rightarrow 0) \rightarrow C$ is inconsistent because the corresponding recursion principle would allow us to construct an element of type $0$.
+  Another example of an inconsistent type for a constructor is $((D \rightarrow Prop) \rightarrow Prop) \rightarrow D)$.
+  To keep our types consistent it suffices to restrict ourselves to constructors whose domains have strict positivity.
+\end{remark}
+
+
+\section{Initial Algebras}
+One way to formalize inductive types is to give them a universal property as initial objects in a category of appropriate ``algebras'', where the ``algebras'' in this category are types with the same structure as the structure imposed by the constructors.
+This ``algebra'' structure will ensure that every object in the category is in some sense ``at most'' the inductive type we want to define.
+To pin down which particular object in this category is the freely generated inductive type, we want to choose the object that has as few additional restrictions and relations as possible.
+This choosing the least restricted (or most free) object will be captured by choosing the initial object.
+
+To make these abstractions more concrete, we'll consider just the inductive type $\N$ and the corresponding category of $\N$-algebras.
+We want objects in this category to be types with the same structure as $\N$.
+Such a type $A$ should be types equipped with some element $zero_A : A$ and some function $suc_A : A \rightarrow A$.
+We can capture this by defining the type of $\N$-algebras to be
+\begin{equation*}
+  \N Alg :\equiv \sum_{A : U} A \times (A \rightarrow A)
+\end{equation*}
+We can construct many examples of $\N-Algebras$ besides $\N$ fairly easily.
+For example ($\2$, $true$, $\lambda x . \neg x$) is an $\N$-algebra.
+More generally, for any inhabited type $C$, there exists some $c : C$, and  ($C$, $c$, $\lambda x . x$) is an $N$-algebra.
+
+Next we define $\N$-homomorphism between two $\N$-algebras $(C, c_0, c_s)$ and $(D, d_0, d_s)$ to be a function $h: C \rightarrow D$ such that $h(c_0) = d_0$ and $h(c_s(c)) = d_s(h(c))$ for all $c : C$. We can pack this into a single type by defining the type of $\N$-homomorphisms to be
+\begin{equation*}
+  \N Hom((C,c_0,c_s), (D,d_0,d_s)) :\equiv \sum_{h : C \rightarrow D}\parens*{(h(c_0)=d_0) \times \prod_{c : C} (h(c_s(c))=d_s(h(c)))}
+\end{equation*}
+
+Then together these form a category whose objects are the $\N$-algebras and whose arrows are the $\N$-homomorphisms.
+Then we claim that $\N$ is \textbf{the} initial object in this category, where an initial object is one that has a unique arrow from it to any other object in the category.
```

Comment:
I glossed over the details in lectures, but to be 500% accurate, the homomorphism is unique up to homotopy. So the initiality is understood as *homotopy* initiality.